src/iop/freaky_bw.c

```diff
@@ -0,0 +1,1217 @@
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+/*
+   based on G'MIC filter "Freaky B&W"
+   http://gmic.eu/
+
+   it also have a dst version in addition to the fft one
+
+#@gui Freaky B&amp;W : fx_freaky_bw, fx_freaky_bw_preview
+#@gui : Strength (%) = float(90,0,100)
+#@gui : Oddness (%) = float(20,0,100)
+#@gui : Brightness (%) = float(0,-100,100)
+#@gui : Contrast (%) = float(0,-100,100)
+#@gui : Gamma (%) = float(0,-100,100)
+#@gui : sep = separator(), Preview type = choice("Full","Forward horizontal","Forward vertical","Backward
+horizontal","Backward vertical","Duplicate top","Duplicate left","Duplicate bottom","Duplicate right")
+#@gui : sep = separator(), note = note("<small>Author: <i>David Tschumperl&#233;</i>.      Latest update:
+<i>09/30/2015</i>.</small>")
+fx_freaky_bw :
+  -repeat $! -l[$>] -split_opacity -l[0]
+    -to_rgb
+
+    # Estimate gradient field of B&W result.
+    --expand_xy 1,0 -channels. 0,4
+    -f. ">if (c!=4,i,
+           Rx = i(x+1,y,0,0) - i(x,y,0,0);
+           Ry = i(x,y+1,0,0) - i(x,y,0,0);
+           Rn = Rx^2 + Ry^2;
+           Gx = i(x+1,y,0,1) - i(x,y,0,1);
+           Gy = i(x,y+1,0,1) - i(x,y,0,1);
+           Gn = Gx^2 + Gy^2;
+           Bx = i(x+1,y,0,2) - i(x,y,0,2);
+           By = i(x,y+1,0,2) - i(x,y,0,2);
+           Bn = Bx^2 + By^2;
+           n = 1e-5 + max(Rn,Gn,Bn)^"{$2%}";
+           val = 0;
+          if (Rn>=Gn && Rn>=Bn,
+            i(x,y,0,3) = Rx/n; val=Ry/n,
+          if (Gn>=Rn && Gn>=Bn,
+            i(x,y,0,3) = Gx/n; val=Gy/n,
+            i(x,y,0,3) = Bx/n; val=By/n));
+          val
+         )"
+    -channels. 3,4
+    -luminance[0] ia={0,ia}
+
+    # Estimate laplacian of final image.
+    -s. c
+    -f.. "i - i(x-1,y,0,0)"
+    -f. "i - i(x,y-1,0,0)"
+    -+[-2,-1]
+
+    # Reconstruct image from laplacian.
+    -fft.
+    100%,100%,1,1,'cos(2*x*pi/w)+cos(2*y*pi/h)' -*. 2 --. 4
+    -=. 1 -/[-3,-2] . -rm.
+    -=.. 0 -=. 0
+    -ifft[-2,-1] -rm.
+    -shrink_xy. 1 -+. $ia -n. 0,255
+
+    # Merge result with original color image.
+    -j[0] [1],0,0,0,0,{$1%} -rm.
+    -adjust_colors ${3-5}
+  -endl -a c -endl -done
+
+fx_freaky_bw_preview :
+  -gui_split_preview "-fx_freaky_bw $*",$-1
+
+*/
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "bauhaus/bauhaus.h"
+#include "common/colorspaces_inline_conversions.h"
+#include "develop/imageop_math.h"
+
+#ifdef HAVE_FFTW3
+#include <fftw3.h>
+#endif
+
+DT_MODULE_INTROSPECTION(1, dt_iop_fbw_params_t)
+
+typedef struct
+{
+  float *in_src;
+  fftwf_complex *out_src;
+
+  fftwf_plan plan_src;
+  fftwf_plan plan_inv;
+
+  int width_src;
+  int height_src;
+
+  int width_dest;
+  int height_dest;
+
+  int width_fft;
+  int height_fft;
+  int width_fft_complex;
+  int height_fft_complex;
+
+  dt_pthread_mutex_t *fftw3_lock;
+} fbw_fft_t;
+
+typedef enum dt_iop_fbw_bw_methods_t {
+  dt_iop_fbw_bw_mix_dst = 0,
```

Comment:
should be all upper-case : DT_IOP_FBW_BW_MIX_DST

Reply:
Sure, but let's first work on the description of the 6 modes, I really don't have a clue what text to use so it makes some sense for the user, not everyone is familiar with the fourier or sine transform and the laplacian, and the user shouldn't even care about it.